targets/esp8266/jshardware.c

```diff
@@ -85,17 +95,65 @@ static void utilTimerInit(void);
  */
 void jshInit() {
   // A call to jshInitDevices is architected as something we have to do.
+  os_printf("> jshInit\n");
+
+  // Initialize the ESP8266 GPIO subsystem.
+  gpio_init();
+
   systemTimeInit();
   utilTimerInit();
   jshInitDevices();
+
+  // Register a callback function to be called for a GPIO interrupt
+  gpio_intr_handler_register(intrHandlerCB, NULL);
+
+  ETS_GPIO_INTR_ENABLE();
+  os_printf("< jshInit\n");
 } // End of jshInit
 
 /**
+ * Handle a GPIO interrupt.
+ * We have arrived in this callback function because the state of a GPIO pin has changed
+ * and it is time to record that change.
+ */
+static void intrHandlerCB(
+    uint32 interruptMask, //!< A mask indicating which GPIOs have changed.
+    void *arg             //!< Optional argument.
+  ) {
+  // Given the interrupt mask, we as if bit "x" is on.  If it is, then that is defined as meaning
+  // that the state of GPIO "x" has changed so we want to raised an event that indicates that this
+  // has happened...
+  // Once we have handled the interrupt flags, we need to acknowledge the interrupts so
+  // that the ESP8266 will once again cause future interrupts to be processed.
+
+  os_printf(">> intrHandlerCB\n");
+  gpio_intr_ack(interruptMask);
+  // We have a mask of interrupts that have happened.  Go through each bit in the mask
+  // and, if it is on, then an interrupt has occurred on the corresponding pin.
+  int pin;
+  for (pin=0; pin<16; pin++) {
+    if ((interruptMask & (1<<pin)) != 0) {
+      // Pin has changed so push the event that says pin has changed.
+      jshPushIOWatchEvent(pinToEV_EXTI(pin));
+      gpio_pin_intr_state_set(GPIO_ID_PIN(pin), GPIO_PIN_INTR_ANYEDGE);
+    }
+  }
+  os_printf("<< intrHandlerCB\n");
+}
+
+/**
  * Reset the Espruino environment.
  */
 void jshReset() {
   //system_restart();
   os_printf("> jshReset\n");
+  // Set all GPIO pins to be input.
+  /*
+  int i;
+  for (int i=0; i<16; i++) {
+    jshPinSetState(i, JSHPINSTATE_GPIO_IN);
```

Comment:
There's a #define around - if you look at the STM32 implementation - that is something like 'is_pin_used' - it stops pins such as the serial IO getting set to inputs